Fix ECDH agree

chfast opened the pull request:
We must consider secret and public key being invalid inputs of ecdh::agree().

Fixes https://github.com/ethereum/cpp-ethereum/issues/4020.

gumb0 commented on libdevcrypto/ECDHE.cpp:
```diff
@@ -31,6 +31,7 @@ void ECDHE::agree(Public const& _remote, Secret& o_sharedSecret) const
 	m_remoteEphemeral = _remote;
-	ecdh::agree(m_ephemeral.secret(), m_remoteEphemeral, o_sharedSecret);
+	if (!ecdh::agree(m_ephemeral.secret(), m_remoteEphemeral, o_sharedSecret))
+		BOOST_THROW_EXCEPTION(InvalidState());
```
You should probably create a new exception type for this, since this looks completely different from the error checked above

chfast replied:
Ok, I will refactor that. Thanks.

chfast commented:
So I removed the ECDHE to get rid of "invalid state" exception there. And I added InvalidPublic exception to inform about invalid pubkey passed for ECDH. Still there is a lot of mix conventions how to handle errors...

chriseth commented on libdevcrypto/ECDHE.cpp:
```diff
@@ -1,37 +0,0 @@
-
-void ECDHE::agree(Public const& _remote, Secret& o_sharedSecret) const
-{
-	if (m_remoteEphemeral)
```
Is this check still in place?

chfast replied:
This check was never effective, because this is agreed during p2p handshake and cannot happen again for the same peer (the same remote public key).

chriseth replied:
I have the feeling that this might be a protection against an attack where you retrieve the private key by performing multiple key agreements. If possible, I would like to keep it as an internal assertion, even (or especially) if it is never triggered by the current code.

chfast replied:
Fixed. And assert included.

chfast commented:
Now I'm f*cked! I did more than requested and p2p is broken now. So I either revert it and apply simple fix requested by @gumb0 or try to fix it :).

chfast commented:
Can we merge it soon?

chfast commented:
I moved the exception to the p2p namespace. The comments also explain where it is caught. The handshake will just fail nicely, what is good.

gumb0 commented on libp2p/RLPxHandshake.h:
```diff
@@ -58,8 +58,6 @@ class RLPXHandshake: public std::enable_shared_from_this<RLPXHandshake>
 	/// Setup outbound connection.
 	RLPXHandshake(Host* _host, std::shared_ptr<RLPXSocket> const& _socket, NodeID _remote): m_host(_host), m_remote(_remote), m_originated(true), m_socket(_socket), m_idleTimer(m_socket->ref().get_io_service()) { crypto::Nonce::get().ref().copyTo(m_nonce.ref()); }
 
-	~RLPXHandshake() {}
```
The class has a virtual protected function, shouldn't it have virtual destructor?

chfast replied:
It has because the base class has. The one I removed was actually hiding the virtual one.

chfast replied:
Changed to virtual. Sorry.
